# Nettoyage des données V2

In [20]:
pd.set_option("display.max_columns", None)

### Récupération des données

In [1]:
from scripts.importation import process_all_years_s3

In [2]:
import s3fs
import pandas as pd

In [3]:
# Connexion au bucket S3
fs = s3fs.S3FileSystem(client_kwargs={"endpoint_url": "https://minio.lab.sspcloud.fr"})

ROOT_S3_PATH = "mvallat/diffusion/projet"

In [4]:
# Importation bucket S3
MY_BUCKET = "mvallat/diffusion"
FILE_PATH_S3 = f"{MY_BUCKET}/projet/data_brevets.parquet"

with fs.open(FILE_PATH_S3, 'rb') as file_in:
    data_brevets = pd.read_parquet(file_in)

data_brevets.head(10)

doc-number kind country  status publication_country publication_doc-number  \
0    3038352   A1      FR  PUBDEM                  FR                3038352   
1    3038278   A1      FR  PUBDEM                  FR                3038278   
2    3038409   A1      FR  PUBDEM                  FR                3038409   
3    3038425   A1      FR  PUBDEM                  FR                3038425   
4    3038316   A1      FR  PUBDEM                  FR                3038316   
5    3038399   A1      FR  PUBDEM                  FR                3038399   
6    3038381   A1      FR  PUBDEM                  FR                3038381   
7    3038328   A1      FR  PUBDEM                  FR                3038328   
8    3038251   A1      FR  PUBDEM                  FR                3038251   
9    3038374   A1      FR  PUBDEM                  FR                3038374   

  publication_date publication_bopinum publication_nature  \
0         20170106             2017-01             Brevet   
1         20170106             2017-01             Brevet   
2         20170106             2017-01             Brevet   
3         20170106             2017-01             Brevet   
4         20170106             2017-01             Brevet   
5         20170106             2017-01             Brevet   
6         20170106             2017-01             Brevet   
7         20170106             2017-01             Brevet   
8         20170106             2017-01             Brevet   
9         20170106             2017-01             Brevet   

                                     invention-title  ... citation_3_country  \
0  ENSEMBLE POUR PANNEAU DE PORTE ET PANNEAU ASSOCIE  ...                 NA   
1  VEHICULE HYBRIDE COMPRENANT UNE SUPPRESSION DE...  ...                 NA   
2  PROCEDE DE CONCEPTION ASSISTEE PAR ORDINATEUR ...  ...                 NA   
3  DOCUMENT ELECTRONIQUE TEL QU'UNE CARTE A PUCE ...  ...                 NA   
4  PROCEDE DE SYNTHESE EN CONTINU D'UN ELASTOMERE...  ...                 NA   
5  PROCEDE ET DISPOSITIF DE COMMANDE D'UNE CONSIG...  ...                 NA   
6  DISPOSITIF CAPTEUR PREVU POUR ETRE EMBARQUE DE...  ...                 NA   
7  VEHICULE FERROVIAIRE POUR LE NETTOYAGE DES FIL...  ...                 NA   
8  PLANCHER PVC EN FIBRES COUPEES ET PROCEDE DE F...  ...                 NA   
9  INSTALLATION THERMIQUE DE SECHAGE DE MATIERE P...  ...                 NA   

  citation_3_doc-number citation_3_date  year application_country  \
0                    NA              NA  2017                None   
1                    NA              NA  2017                None   
2                    NA              NA  2017                None   
3                    NA              NA  2017                None   
4                    NA              NA  2017                None   
5                    NA              NA  2017                None   
6                    NA              NA  2017                None   
7                    NA              NA  2017                None   
8                    NA              NA  2017                None   
9                    NA              NA  2017                None   

  application_doc-number application_date last-fee-payement next-fee-payement  \
0                   None             None              None              None   
1                   None             None              None              None   
2                   None             None              None              None   
3                   None             None              None              None   
4                   None             None              None              None   
5                   None             None              None              None   
6                   None             None              None              None   
7                   None             None              None              None   
8                   None             None              None              None  

### Observation des variables

#### **1) Type de brevet**

In [5]:
# Colonne "kind"
data_brevets["kind"].value_counts()

kind
A1    107749
A3      4441
NA       102
B1         3
Name: count, dtype: int64

Le document contient trois types de brevets :

- **A1** : demande de brevet  
- **A3** : demande de certificat d'utilité  
- **B1** : brevet délivré  


**Concernant les types B1 (n = 3)** : cette base recensant les demandes de brevet, elle n'est pas censée contenir un brevet B1.  
Probablement dû à une erreur de classification, nous pouvons donc retirer ces observations.

In [23]:
# Suppression des B1
data_brevets_clean = data_brevets[data_brevets["kind"] != "B1"]

# Vérification qu'on a supprimé 3 observations
data_brevets_clean.shape[0]


112292

**Concernant les types A3 (n = 4 441)** : les certificats d'utilité sont des certificats délivrés par l'INPI qui protègent l'innovation, tout comme les brevets. En revanche la durée de protection est raccourcie (10 ans contre 20 ans) et la procédure d'examen de la demande est allégée. Pour simplifier l'analyse en la rendant plus homogène, nous ne gardons pas ces demandes A3 dans notre base.

In [24]:
# Suppression des A3
data_brevets_clean = data_brevets_clean[data_brevets_clean["kind"] != "A3"]

# Vérification : suppression de 4 441 observations (112,292 - 4,441 = 107,851)
data_brevets_clean.shape[0]

107851

**Concernant les NA (n = 102)** : plus d'observations sont nécessaires pour décider de la manière de gérer ces NA.


In [25]:
# Observation des NA
brevets_na = data_brevets[data_brevets["kind"] == "NA"]
brevets_na


doc-number kind country status publication_country  \
285            NA   NA      NA     NA                None   
606            NA   NA      NA     NA                None   
831            NA   NA      NA     NA                None   
1153           NA   NA      NA     NA                None   
1543           NA   NA      NA     NA                None   
...           ...  ...     ...    ...                 ...   
28236          NA   NA      NA     NA                None   
28556          NA   NA      NA     NA                None   
28760          NA   NA      NA     NA                None   
29096          NA   NA      NA     NA                None   
107329    3145430   NA      FR    NEW                  FR   

       publication_doc-number publication_date publication_bopinum  \
285                      None             None                None   
606                      None             None                None   
831                      None             None                None   
1153                     None             None                None   
1543                     None             None                None   
...                       ...              ...                 ...   
28236                    None             None                None   
28556                    None             None                None   
28760                    None             None                None   
29096                    None             None                None   
107329                3145430         20240802             2024-31   

       publication_nature                                    invention-title  \
285                  None                                                 NA   
606                  None                                                 NA   
831                  None                                                 NA   
1153                 None                                                 NA   
1543                 None                                                 NA   
...                   ...                                                ...   
28236                None                                                 NA   
28556                None                                                 NA   
28760                None                                                 NA   
29096                None                                                 NA   
107329  Demande française  Surveillance de l’intégrité des données et con...   

       applicant_1_orgname     applicant_1_address-1 applicant_1_city  \
285                     NA                        NA               NA   
606                     NA                        NA               NA   
831                     NA                        NA               NA   
1153                    NA                        NA               NA   
1543                    NA                        NA               NA   
...                    ...                       ...              ...   
28236                   NA                        NA               NA   
28556                   NA                        NA               NA   
28760                   NA                        NA               NA   
29096                   NA                        NA               NA   
107329             AMADEUS  485 ROUTE DU PIN MONTARD             BIOT   

       applicant_1_postcode applicant_1_country applicant_2_orgname  \
285                      NA                  NA                  NA   
606                      NA                  NA                  NA   
831                      NA                  NA                  NA   
1153                     NA                  NA                  NA   
1543                     NA                  NA                  NA   
...                     ...                 ...                 ...   
28236                    NA                  NA                  NA   
28556                    NA

In [26]:
# Vérification du nombre d'informations réellement présentes
brevets_na["country"].nunique()
brevets_na["invention-title"].nunique()

2

On peut constater que seules deux observations contiennent des informations. Nous pouvons donc supprimer toutes les observations qui ne contiennent que des NA.

In [27]:
# Suppression des observations 100% NA
data_brevets_clean = data_brevets_clean[data_brevets_clean["kind"] != "NA"]

#### **2) Numéro du document**

In [28]:
# Vérification que doc_number est unique

nb_number = data_brevets_clean["doc-number"].nunique()
print("Nombre de doc-number uniques :", nb_number)

nb_number2 = data_brevets_clean["publication_doc-number"].nunique()
print("Nombre de publication_doc-number uniques :", nb_number2)

nb_observations = data_brevets_clean.shape[0]
print("Nombre total d'observations :", nb_observations)


Nombre de doc-number uniques : 107748
Nombre de publication_doc-number uniques : 107748
Nombre total d'observations : 107749


Nous pouvons constater qu'il y a une observation qui n'a pas un numéro unique.

In [29]:
doublons = data_brevets_clean[data_brevets_clean.duplicated(subset=["doc-number"], keep=False)]
print(doublons)

      doc-number kind country  status publication_country  \
12476    3050837   A1      FR  PUBDEM                  FR   
58919    3050837   A1      FR     NEW                  FR   

      publication_doc-number publication_date publication_bopinum  \
12476                3050837         20171103             2017-44   
58919                3050837         20171103             2017-44   

      publication_nature                                    invention-title  \
12476             Brevet  DISPOSITIF D'INFORMATION INERTIELLE, SAGITTALE...   
58919  Demande française  DISPOSITIF D'INFORMATION INERTIELLE, SAGITTALE...   

                  applicant_1_orgname      applicant_1_address-1  \
12476  BOARDING RING READING SAS None  661 CHE DU CHATEAU VALLON   
58919       BOARDING RING READING SAS  661 CHE DU CHATEAU VALLON   

      applicant_1_city applicant_1_postcode applicant_1_country  \
12476        OLLIOULES                83190                  FR   
58919        OLLIOULES         

#### **3) Statut**

La documentation de l'INPI nous apprend que la variable "status" est utilisée pour indiquer quel est l'évènement qui a entraîné la présence du brevet ou du CCP dans le fichier de la semaine. Les 

In [13]:
# Colonne "status"
data_brevets_clean["status"].value_counts()

status
NEW        79065
PUBDEM     28124
AMENDED      560
Name: count, dtype: int64

In [14]:
data_brevets_clean.shape[0]

107749

#### **4) Abstract**
Les abstracts sont les résumés des brevets déposés. Il s'agit pour nous de vérifier qu'ils contiennent bien du texte d'une part, et qu'il n'y a pas de doublons d'autre part.

In [22]:
counts = data_brevets_clean["abstract"].value_counts()
print(counts[counts > 1])

abstract
NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             

On peut constater qu'il y a 1166 abstracts en double. On peut tout d'abord supprimer ceux qui contiennent des NA car il semble probable que cela soit du à une erreur dans le dossier original, une demande de brevet sans abstract n'étant pas recevable. Il y en a 339, ce qui n'est pas un nombre trop conséquent au vu de nos plus de 100,000 observations.

In [30]:
data_brevets_clean.groupby("abstract").filter(lambda x: len(x) > 1)


doc-number kind country  status publication_country  \
48        3038286   A1      FR  PUBDEM                  FR   
76        3038249   A1      FR  PUBDEM                  FR   
84        3038325   A1      FR  PUBDEM                  FR   
131       3038279   A1      FR  PUBDEM                  FR   
160       3038242   A1      FR  PUBDEM                  FR   
...           ...  ...     ...     ...                 ...   
112223    3150324   A1      FR     NEW                  FR   
112224    3150325   A1      FR     NEW                  FR   
112225    3150326   A1      FR     NEW                  FR   
112288    3150389   A1      FR     NEW                  FR   
112289    3150390   A1      FR     NEW                  FR   

       publication_doc-number publication_date publication_bopinum  \
48                    3038286         20170106             2017-01   
76                    3038249         20170106             2017-01   
84                    3038325         20170106             2017-01   
131                   3038279         20170106             2017-01   
160                   3038242         20170106             2017-01   
...                       ...              ...                 ...   
112223                3150324         20241227             2024-52   
112224                3150325         20241227             2024-52   
112225                3150326         20241227             2024-52   
112288                3150389         20241227             2024-52   
112289                3150390         20241227             2024-52   

       publication_nature                                    invention-title  \
48                 Brevet  ENSEMBLE COMPRENANT UN GARDE-BOUE POUR VEHICUL...   
76                 Brevet  VITRAGE CHAUFFANT A FEUILLE DE VERRE EXTERIEUR...   
84                 Brevet  PROCEDE DE CRYOCONSERVATION DE CELLULES A VISE...   
131                Brevet  PROCEDE DE CONDUITE AUTOMATIQUE POUR L'INSERTI...   
160                Brevet  ALLIAGE D'ALUMINIUM POUR SOUDAGE PAR LASER SAN...   
...                   ...                                                ...   
112223  Demande française  Procédé de traitement informatique de données ...   
112224  Demande française  Procédé de traitement informatique de données ...   
112225  Demande française  Procédé de traitement informatique de données ...   
112288  Demande française  Dispositif de refroidissement configuré pour r...   
112289  Demande française  Dispositif de refroidissement configuré pour r...   

                                      applicant_1_orgname  \
48                    PEUGEOT CITROEN AUTOMOBILES SA None   
76                         SAINT-GOBAIN GLASS FRANCE None   
84      LABORATOIRE FRANCAIS DU FRACTIONNEMENT ET DES ...   
131     COMMISSARIAT A L'ENERGIE ATOMIQUE ET AUX ENERG...   
160                         CONSTELLIUM NEUF-BRISACH None   
...                                                   ...   
112223                                       Rayon Thomas   
112224                                       Rayon Thomas   
112225                                       Rayon Thomas   
112288                          VALEO SYSTEMES THERMIQUES   
112289                          VALEO SYSTEMES THERMIQUES   

                                    applicant_1_address-1  \
48                                          ROUTE DE GISY   
76                                     18 AVENUE D'ALSACE   
84                   ZA DE COURTABOEUF 3 AV DES TROPIQUES   
131              BATIMENT "LE PONANT D", \n25 RUE LEBLANC   
160                               ZIP RHENANE NORD\nRD 52   
...                                                   ...   
112223  Entrée Boutique Coup d'oeil 6 Rue Joseph Marie...   
112224  Entrée Boutique Coup d'oeil 6 Rue Joseph Marie...   
112225  Entrée Boutique Coup d'oeil 6 Rue Joseph Marie...   
112288                    8 rue Louis Lormand La Verrière   
112289                    8 rue Louis Lormand La Verrièr

In [16]:
data_brevets_clean["abstract"].nunique()


105885

### Tri des variables pertinentes

In [17]:
data_brevets_light = data_brevets[['publication_date', 'invention-title', 'classification_1_text', 'abstract']]
data_brevets_clean.head(5)


doc-number kind country  status publication_country publication_doc-number  \
0    3038352   A1      FR  PUBDEM                  FR                3038352   
1    3038278   A1      FR  PUBDEM                  FR                3038278   
2    3038409   A1      FR  PUBDEM                  FR                3038409   
3    3038425   A1      FR  PUBDEM                  FR                3038425   
4    3038316   A1      FR  PUBDEM                  FR                3038316   

  publication_date publication_bopinum publication_nature  \
0         20170106             2017-01             Brevet   
1         20170106             2017-01             Brevet   
2         20170106             2017-01             Brevet   
3         20170106             2017-01             Brevet   
4         20170106             2017-01             Brevet   

                                     invention-title  ... citation_3_country  \
0  ENSEMBLE POUR PANNEAU DE PORTE ET PANNEAU ASSOCIE  ...                 NA   
1  VEHICULE HYBRIDE COMPRENANT UNE SUPPRESSION DE...  ...                 NA   
2  PROCEDE DE CONCEPTION ASSISTEE PAR ORDINATEUR ...  ...                 NA   
3  DOCUMENT ELECTRONIQUE TEL QU'UNE CARTE A PUCE ...  ...                 NA   
4  PROCEDE DE SYNTHESE EN CONTINU D'UN ELASTOMERE...  ...                 NA   

  citation_3_doc-number citation_3_date  year application_country  \
0                    NA              NA  2017                None   
1                    NA              NA  2017                None   
2                    NA              NA  2017                None   
3                    NA              NA  2017                None   
4                    NA              NA  2017                None   

  application_doc-number application_date last-fee-payement next-fee-payement  \
0                   None             None              None              None   
1                   None             None              None              None   
2                   None             None              None              None   
3                   None             None              None              None   
4                   None             None              None              None   

  date-search-completed  
0                  None  
1                  None  
2                  None  
3                  None  
4                  None  

[5 rows x 102 columns]